src/main/java/ch/njol/skript/SkriptAddon.java

```diff
@@ -21,123 +21,195 @@
-			v = new Version("" + p.getDescription().getVersion());
+			version = new Version(descriptionVersion);
 		} catch (final IllegalArgumentException e) {
-			final Matcher m = Pattern.compile("(\\d+)(?:\\.(\\d+)(?:\\.(\\d+))?)?").matcher(p.getDescription().getVersion());
+			Matcher m = Pattern.compile("(\\d+)(?:\\.(\\d+)(?:\\.(\\d+))?)?").matcher(descriptionVersion);
 			if (!m.find())
-				throw new IllegalArgumentException("The version of the plugin " + p.getName() + " does not contain any numbers: " + p.getDescription().getVersion());
-			v = new Version(Utils.parseInt("" + m.group(1)), m.group(2) == null ? 0 : Utils.parseInt("" + m.group(2)), m.group(3) == null ? 0 : Utils.parseInt("" + m.group(3)));
-			Skript.warning("The plugin " + p.getName() + " uses a non-standard version syntax: '" + p.getDescription().getVersion() + "'. Skript will use " + v + " instead.");
+				throw new IllegalArgumentException("The version of the plugin " + plugin.getName() + " does not contain any numbers: " + descriptionVersion);
+			version = new Version(Utils.parseInt("" + m.group(1)), m.group(2) == null ? 0 : Utils.parseInt("" + m.group(2)), m.group(3) == null ? 0 : Utils.parseInt("" + m.group(3)));
+			Skript.warning("The plugin " + plugin.getName() + " uses a non-standard version syntax: '" + descriptionVersion + "'. Skript will use " + version + " instead.");
 		}
-		version = v;
+		this.version = version;
 	}
 	
 	@Override
-	public final String toString() {
-		return name;
+	public String toString() {
+		return plugin.getName();
 	}
 	
 	public String getName() {
-		return name;
+		return plugin.getName();
 	}
-	
+
 	/**
-	 * Loads classes of the plugin by package. Useful for registering many syntax elements like Skript does it.
-	 * 
-	 * @param basePackage The base package to add to all sub packages, e.g. <tt>"ch.njol.skript"</tt>.
-	 * @param subPackages Which subpackages of the base package should be loaded, e.g. <tt>"expressions", "conditions", "effects"</tt>. Subpackages of these packages will be loaded
-	 *            as well. Use an empty array to load all subpackages of the base package.
-	 * @throws IOException If some error occurred attempting to read the plugin's jar file.
+	 * Loads classes of the plugin by package. Useful for registering many syntax elements like Skript.
+	 *
+	 * Please note that if you need to load the same class multiple times,
+	 * you should call {@link #resetEntryCache()} each time you call this method.
+	 *
+	 * @param basePackage The base package to start searching in (e.g. 'ch.njol.skript').
+	 * @param subPackages Specific subpackages to search in (e.g. 'conditions')
+	 *                    If no subpackages are provided, all subpackages of the base package will be searched.
+	 * @return This SkriptAddon.
+	 */
+	public SkriptAddon loadClasses(String basePackage, String... subPackages) {
+		return loadClasses(null, true, basePackage, true, subPackages);
+	}
+
+	@Nullable
+	private JarEntry @Nullable [] entryCache;
```

Comment:
The fields are grouped with their own methods (see `file` or `languageFileDirectory`), which I think can be useful in organizing the classes :D

Reply:
Contents for this array will no longer be null, so I have removed the annotation for the array's contents

In case anyone is not aware, to annotate the array itself as nullable the annotation must be placed after the type declaration but before the brackets